### PR TITLE
Fixes maps wms zoom limitation

### PR DESCRIPTION
**Pull request by junqiu-lei**
Signed-off-by: Junqiu Lei <junqiu@amazon.com>

### Description
When using WMS to customize base tile maps in region map or coordinate maps, user aren’t able to get into higher zoom levels than opensearch maps service default maximum zoom level.

By this PR fixed, the WMS max zoom value can be reached to 22 which is originally deigned [here](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/9678a4bed88361574dc75f872c31d68834986562/src/plugins/maps_legacy/public/map/base_maps_visualization.js#L39)
 
### Issues Resolved
https://github.com/opensearch-project/maps/issues/19
 
### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
    - [X] `yarn test:jest`
    - [X] `yarn test:jest_integration`
    - [X] `yarn test:ftr`
- [ ] New functionality has been documented.
- [X] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/maps_legacy/public/map/base_maps_visualization.js [hunk 1]
```diff
@@ -164,7 +164,7 @@ export function BaseMapsVisualizationProvider() {
 
       try {
         if (this._wmsConfigured()) {
-          if (WMS_MINZOOM > this._opensearchDashboardsMap.getMaxZoomLevel()) {
+          if (WMS_MAXZOOM > this._opensearchDashboardsMap.getMaxZoomLevel()) {
```
Comment (vamshin) [inline src/plugins/maps_legacy/public/map/base_maps_visualization.js:167]: Good catch! Thanks for the fix.

**Review by vamshin — DISMISSED** [review]
LGTM! Thanks

**Review by martin-gaievski — DISMISSED** [review]

**Comment by martin-gaievski** [comment]
Do we have unit/any other type of tests in this project? If so I think we need to update tests and add checks for edge cases like min/max values

**Comment by junqiu-lei** [comment]
> Do we have unit/any other type of tests in this project? If so I think we need to update tests and add checks for edge cases like min/max values

I didn't find any existing test file for `base_maps_visualization.js` to add for this PR. I have tested it manually on region maps and coordinate maps when at wms or default maps.

**Review by ashwin-pc — DISMISSED** [review]
LGTM

**Comment by joshuarrrr** [comment]
> Do we have unit/any other type of tests in this project? If so I think we need to update tests and add checks for edge cases like min/max values

@martin-gaievski We do have functional tests that test min/max zoom levels - see https://github.com/opensearch-project/OpenSearch-Dashboards/pull/1823/files, for example.

File: src/plugins/maps_legacy/public/map/base_maps_visualization.js (hunk 1 above)
Comment (joshuarrrr) [reply src/plugins/maps_legacy/public/map/base_maps_visualization.js:167]: Why do we need to also reset the min zoom level in this case (L168)? It seems like maybe the min and max zoom levels should each be reset under different conditions.

**Comment by junqiu-lei** [comment]
> > Do we have unit/any other type of tests in this project? If so I think we need to update tests and add checks for edge cases like min/max values
> 
> @martin-gaievski We do have functional tests that test min/max zoom levels - see https://github.com/opensearch-project/OpenSearch-Dashboards/pull/1823/files, for example.

Hi Josh, thanks for catching, I think the tests here are for default maps tiles zoom levels, not includes WMS, let me check if we can add the WMS zoom test here.

File: src/plugins/maps_legacy/public/map/base_maps_visualization.js (hunk 1 above)
Comment (kavilla) [reply src/plugins/maps_legacy/public/map/base_maps_visualization.js:167]: Could we get more insight the original logic here and the original intent and ensure a regression isn't happening by changing this? Seems like code that has existed since even before the legacy version 6.x and it's changing the value.

From my understanding the original code was checking if the max zoom level was less than 0 then it would reset the min and max to 0 and 22.

Now the logic is if 22 > than max zoom level then it will always set the max zoom level to 22. So it seems regardless of what the map is configure to it will always set the max zoom level to 22.

**Review by kavilla — CHANGES_REQUESTED** [review]

File: src/plugins/maps_legacy/public/map/base_maps_visualization.js (hunk 1 above)
Comment (junqiu-lei) [reply src/plugins/maps_legacy/public/map/base_maps_visualization.js:167]: Hi @kavilla, There was a same issue reported at Kibana 7.1.1 at 2019 https://github.com/elastic/kibana/issues/49709. From that issue context, we could see the WMS zoom level limitation wasn't correct in some versions and didn't get fixed.

Based on this previous PR on Kibana https://github.com/elastic/kibana/commit/4e2401c216064ddc5fd9e7f5161db3e463b6c00a#diff-8eef4c226bf8f716c6420fcec4400202088971b418cb9bc31bdc4cecb57bc365R117, I still think it could be checking  with max zoom level with current existing max zoom level which is from opensearch maps service max zoom level.

If we don't set up the WMS_MAXZOOM to 22, the maps max zoom  would be set and limited by default maps service max zoom. So in order to make sure all WMS users are able to load their WMS tiles at max potential zoom level, we'd better set up  WMS_MAXZOOM to 22.

Hi @joshuarrrr, the code logic already exist since kibana, in my understand, here for WMS cases, it can make sure users always have zoom level from 0 - 22.

**Comment by junqiu-lei** [comment]
Hi @kavilla, how do you think based on my last comment? We can have a offline quick sync about the WMS part code from Kibana if you think it's needed. Then I can update the PR with functional tests coordinately.

File: src/plugins/maps_legacy/public/map/base_maps_visualization.js (hunk 1 above)
Comment (kavilla) [reply src/plugins/maps_legacy/public/map/base_maps_visualization.js:167]: Hi @junqiu-lei, yeah the bug makes sense in that it checks the maps service and defines the max zoom level which doesn't make sense if WMS is defined and it should use WMS configurations.

But I think this logic is intentional to be the case somehow that value returns as a negative and then it has some default case that at least set the same level from a valid (non-negative) value. But with change, regardless what the user configures for max zoom we will always set max zoom to 22. This seems like it addresses the bug as much is the increases the maximum to 22 instead of 10. It also prevents the edge case where that max zoom is a negative value. If somehow the value returned is negative for max zoom then max zoom would be negative.

So is the bug actually occurring at another location? Because I see earlier in the file the map gets configured with these default values and it must be overwritten, but in the rest of this function it sets the options that were configured so perhaps that value is not being applied or the previous base layer needs to be nulled out for WMS to take those settings?

Comment (junqiu-lei) [reply src/plugins/maps_legacy/public/map/base_maps_visualization.js]: Hi @kavilla, I updated the WMS zoom check condition to include both existing check and bug fix check, so that we can  fix the bug and avoid abandon the existing logic.

File: src/plugins/maps_legacy/public/map/base_maps_visualization.js [hunk 2]
```diff
@@ -164,7 +164,8 @@ export function BaseMapsVisualizationProvider() {
 
       try {
         if (this._wmsConfigured()) {
-          if (WMS_MINZOOM > this._opensearchDashboardsMap.getMaxZoomLevel()) {
+          const curMaxZoomLevel = this._opensearchDashboardsMap.getMaxZoomLevel();
```
Comment (kavilla) [inline src/plugins/maps_legacy/public/map/base_maps_visualization.js:167]: current or cur

Comment (junqiu-lei) [reply src/plugins/maps_legacy/public/map/base_maps_visualization.js]: updated to current

**Review by kavilla — APPROVED** [review]
Thanks!

**Review by ashwin-pc — APPROVED** [review]
